### PR TITLE
Blaze: Integrate fetching impressions on the budget setting screen

### LINKED ISSUE #11599: Edit budget screen
image

```[tasklist]
### Tasks
- [ ] Main screen UI
- [ ] Impression explanation
- [ ] Duration setting screen
```

### PR DESCRIPTION
**Pull request by itsmeichigo**
Closes: #11599 

## Description

This PR wraps up the budget settings screen by integrating the fetch for estimated impressions. Changes include:
- Updated `BlazeForecastedImpressionsInput` with `totalBudget` being Double instead of String.
- Added an enum `ForecastedImpressionState` and updated `BlazeBudgetSettingView` following the state.
- Updated `BlazeBudgetSettingViewModel` to observe changes in daily budget, start date and day count to fetch estimated impressions.
- Updated `BlazeCampaignCreationFormViewModel` to send site ID and target options to `BlazeBudgetSettingViewModel`.

## Testing instructions

- Log in to a store with at least one existing product and is eligible for Blaze.
- Navigate to the Products tab and select any public product.
- On the product form, select Promote with Blaze.
- Notice that the campaign creation form is presented. Select the budget field.
- Since the response for forecasted impressions are mocked at the moment, so you'll see the mocked data displayed for estimated impressions.
- Feel free to update the `updateImpressions` method in `BlazeBudgetSettingViewModel` to set the state to error or loading to confirm that the UI is updated accordingly.

## Screenshots

 📝 I improvised the loading and error states. I added a comment on Figma regarding the error state, will update later when I get a response from Ann.
| Loading state | Result state | Error state |
| ----- | ----- | ----- |
|  |  |  | 

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Comment by itsmeichigo** [comment]
Converted the PR to draft to add unit tests.

**Comment by hafizrahman** [comment]
Waiting until this becomes ready for review again. In the meanwhile:

Given that these two sentences are fully tappable, I wonder if we should give them more space to have bigger tap target. But, I'm also aware this might be a temporary design:

Screenshot 2024-01-05 at 08 24 43

**Comment by itsmeichigo** [comment]
> Given that these two sentences are fully tappable, I wonder if we should give them more space to have bigger tap target. 

@hafizrahman Good point! I decided to hide the impression info when fetching fails, that way we would not have to deal with button sizes (also because I could not find a better solution). This PR is now ready for review, please take another look when you can.

File: WooCommerce/Classes/ViewRelated/Blaze/BudgetSetting/BlazeBudgetSettingViewModel.swift [hunk 1]
```diff
@@ -39,33 +38,110 @@ final class BlazeBudgetSettingViewModel: ObservableObject {
...
+        }
+    }
 }
 
+// MARK: - Private helpers
+private extension BlazeBudgetSettingViewModel {
+
+    func observeSettings() {
+        settingSubscription = $dayCount.combineLatest($dailyAmount, $startDate)
+            .debounce(for: .milliseconds(300), scheduler: DispatchQueue.main)
```
Comment (hafizrahman) [inline WooCommerce/Classes/ViewRelated/Blaze/BudgetSetting/BlazeBudgetSettingViewModel.swift:112]: Nice, I was wondering how to best prevent calling impressions value too frequently given that a slider can be moved rather quickly.

File: Networking/Networking/Remote/BlazeRemote.swift [hunk 2]
```diff
@@ -210,34 +210,34 @@ private extension BlazeRemote {
...
         case endDate
-        case formattedTotalBudget = "total_budget"
+        case totalBudget
         case targetings
     }
 }
 
 /// Blaze Forecasted Impressions sub-input related to targetings.
-public struct BlazeTargetOptions: Codable, GeneratedFakeable, GeneratedCopiable {
+public struct BlazeTargetOptions: Encodable, GeneratedFakeable, GeneratedCopiable, Equatable {
```
Comment (hafizrahman) [inline Networking/Networking/Remote/BlazeRemote.swift:240]: Nice catch with the Codable typo.

File: WooCommerce/Classes/ViewRelated/Blaze/CampaignCreation/BlazeCampaignCreationFormViewModel.swift [hunk 3]
```diff
@@ -74,5 +91,11 @@ private extension BlazeCampaignCreationFormViewModel {
             comment: "Blaze campaign budget details with duration in plural form. " +
             "Reads like: $35, 15 days from Dec 31"
         )
+        static let totalBudget = NSLocalizedString(
+            "blazeCampaignCreationFormViewModel.totalBudget",
+            value: "$%.0f USD",
+            comment: "The formatted total budget for a Blaze campaign, fixed in USD. " +
+            "Reads as $11 USD. Keep %.0f as is."
```
Comment (hafizrahman) [inline WooCommerce/Classes/ViewRelated/Blaze/CampaignCreation/BlazeCampaignCreationFormViewModel.swift:98]: Nice that you mentioned the "as is" thing. TIL and something easy to miss.

**Review by hafizrahman — APPROVED** [review]
This looks good code and design wise to me. 👍🏼

**Pull request merged**